### ISSUE TITLE
Accessibility enhancements for mplsjrdevs.com

### ISSUE THREAD

**Issue by jasonwebb** [mplsjrdevs/mplsjrdevs#96]
Hey all! So Andrea Edstrom's talk from February inspired me to help contribute some accessibility upgrades for the mplsjrdevs.com website. I'm a former UI dev/FED, now Developer Advocate over at Accessible360, and accessibility + open-source is kind of my jam. I get to work with a rad team of a11y pros, including experts who are blind or have low vision, and they teach me interesting stuff practically every day that I am more than happy to share with you all 😀

In this thread I'd like to propose a list of a11y improvements for the website, talk through the process I used to identify them, and answer any questions anybody might have. If you'd like to get involved just pop a comment on this thread and we'll figure out a way to put you to work!

## Testing methods

First I ran WebAIM's [WAVE](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh), Google DevTool's [Lighthouse](https://developers.google.com/web/tools/lighthouse#devtools), and Deque's [axe](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) (see results screenshots below). However, since automated tools can't catch everything ([even the best can only cover about ~40%](https://alphagov.github.io/accessibility-tool-audit/)), I also went through the site the using my keyboard (using `Tab`) and the most popular screen readers (NVDA and JAWS).

All of these tests were done on Chrome on a Win10 machine, which lines up well with what the [majority of real-world screen reader users use](https://webaim.org/projects/screenreadersurvey8/#os).

| WAVE | Lighthouse | axe |
|---   |---         |---  |
| ![Screenshot of WAVE results showing 14 issues and 2 warnings](https://user-images.githubusercontent.com/422783/74884642-455a2b80-5339-11ea-97b2-bae63c41e3b2.PNG) | ![Screenshot of Lighthouse results showing a score of 83](https://user-images.githubusercontent.com/422783/74884595-29568a00-5339-11ea-8563-2a60dd1b74b6.PNG) | ![Screenshot of axe results showing 40 errors](https://user-images.githubusercontent.com/422783/74884556-16dc5080-5339-11ea-9066-854f0d1a232d.PNG) |

## Next steps

1. Refine the list for clarity - if you have any questions, ask away!
1. Integrate any suggestions from others that may have been missed (just post to this thread).
1. Decide on branching strategy - a single "accessibility" feature branch, or multiple branches?
1. Break the list up into smaller, more granular Issues that can be discussed independently (maybe one per "area", like "color contrast" or "community carousel")?
1. Assign ownership of tasks to anyone who wants to volunteer
1. Begin working on tasks and opening small, easy-to-review PRs.

## Overview of improvements
Since I'm a whiteboard-and-sticky-notes kinda guy, I created a visual reference in addition to the list of tasks below. Each of these issues are captured in the full list below, I just feel like this is easier to skim for people who don't have the benefit of a super-fast screen reader 😀

<img src="https://user-images.githubusercontent.com/422783/74884695-628efa00-5339-11ea-8468-090cad26e6ed.png" alt="full-page screenshot of mplsjrdevs.com with annotations pointing out locations of all the issues below" height="500px">

## Tasks
I know this seems like a lot, but trust me, its not! This list is only long because I've tried to break up issues into their simplest, most atomic tasks so they can be tackled in an ad-hoc way by people with different experience levels. A lot of them are just adding or removing an attribute, or wrapping some content in an HTML tag. I've also tried to add some minimal explanations to a lot of the items for context, which adds to the overall length.

### Miscellaneous
  - [ ] Add an `<h1>` tag. 
    - Wrapping the existing header logo in an `<h1>` is OK, though some visible text next to it would be awesomer.
  - [ ] Add `tabindex="-1"` to each `<section>` so that the keyboard focus/virtual cursor moves reliably when the same-page nav links are activated.
  - [ ] Wrap the main content sections in a `<main>` landmark tag for easier navigation by screen readers.
    * Make the `#welcome`, `#about`, `#events`, `#community`, and `#code-of-conduct` `<section>`s child elements of this tag. The `#contact` section can be made a `<footer>` (see below).
  - [ ] Add a ["skip to main content"](http://web-accessibility.carnegiemuseums.org/code/skip-link/) link for people using screen readers or only their keyboard. See [WCAG 2.0 SC 2.4.1: Bypass Blocks](http://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-skip.html).
    - This can link to the `<main>` tag created in the previous task.

### Main navigation bar - [see WAI's great tutorial on Menu Structure](https://www.w3.org/WAI/tutorials/menus/structure/)
  - [ ] Wrap the navigation links in a `<nav>` landmark element
  - [ ] Wrap the icon links in a `<ul>` (with individual `<li>`s for the icon links).
  - [ ] Remove the `tabindex` attributes from each icon link. Links with `href`s are natively focusable.
  - [ ] Remove the inner `<button>`s inside each link.
  - [ ] Get the discernable text of the main nav button links to match the text in their tooltips.
    - Right now the `<title>` tags of each SVG icon are being read out by screen readers, which don't quite match their purpose or the tooltip text. If these tags can't be changed, add `aria-hidden="true"` to the SVGs and add sr-only text.

### Mobile menu
  - [ ] Add `aria-hidden="true"` to the icons in the menu.
  - [ ] Add `aria-expanded="true"` to the trigger button, and keep its value updated based on the visibility state of the menu.

### "MN Tech & Diversity Pledge" badge
  - [ ] Add accessible text. Right now this a background image of text, which screen readers can't read, so screen readers skip it entirely.
    - Either make this an `<img>` tag with an appropriate `alt` attribute, or add the text as screen-reader-only text within the `<div class="diversitybadge">` element.
  - [ ] Move the DOM element for the badge (`<div class="diversitybadge">`) to a location that is similar to it's initial visual location. 
    - Right now its placed at the end of the DOM, which nearly every screen reader user will not encounter because it comes after the footer (which they expect to be the last element). I recommend putting it right after the `<header>`, maybe even inside of the "Welcome" section.
  - [ ] _(UX suggestion)_ Can this be wrapped in a link to the pledge or a section talking about _how_ Mpls Jr. Devs is implementing the pledge? I'm not sure what this pledge actually is, but I'd like to learn more!

### Improve or add focus indicators for all interactive elements - [see WCAG 2.0 SC 2.4.7: Focus Visible](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-visible.html)
  - [ ] Main navigation menu icon links. The browser defaults don't stand out very well.
  - [ ] Buttons under Events, Code of Contact, and Contact sections (the outlines are either invisible or very faint)
  - [ ] Links under the Events section (like author website links)
  - [ ] Social media icon links in Community carousel content
  - [ ] Bio thumbnails in Community carousel
  - [ ] Text links in the Contact section.
  - [ ] Mobile menu trigger button

### Color contrast - [see WCAG 2.0 SC 1.4.3: Contrast (Minimum)](http://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html)
  - [ ] Headings (`<h3>`s) "Upcoming" and "Past Events" in the Events section
  - [ ] Gray-ish text for each event (`#event .text`)
  - [ ] Pink buttons with white text (`Events > Show More Events`, `Code of Conduct > View Code of Conduct`, and `Footer > Subscribe`)

### Events section
  - [ ] Add `[aria-disabled="true"]` to the visually-disabled button under Events that says "Registration opens Feb 25th"
    - Consider reducing ambiguity by converting this to static text instead of a `<button>`. When people using screen readers (or their keyboard) reach this button, they expect it to do something when its activated, and could be confused when nothing happens. 

### Community carousel
  - [ ] Miscellaneous
    - [ ] Wrap the entire carousel in a `<section>` with an `aria-label="community carousel"` attribute. 
      - This will cause screen readers to announce when their users enter and leave the carousel, and even skip it entirely if they want. 
    - [ ] Add some clear, succinct screen-reader-only text as the first child element of the `<section>` (above) describing how the carousel works. 
      - For example, "this is a carousel with auto-rotating slides. Activate any of the buttons to disable rotation and 
    - [ ] _(UX suggestion)_ Consider EITHER placing the thumbnails before the bios (on the left for desktop, top for mobile) OR stating in the screen-reader-only instructions that the controls follow the content
  - [ ] Controls and behaviors
    - [ ] Add a pause button as the first focusable element. See [WCAG 2.0 2.2.2: Pause, Stop, Hide](http://www.w3.org/TR/UNDERSTANDING-WCAG20/time-limits-pause.html) for more.
    - [ ] Automatically pause auto-rotation when any of the thumbnails are activated so that the bio doesn't unexpectedly change while the user is trying to get back to the content panel from the thumbnails.
  - [ ] Bio section
    - [ ] Add accessible text to the social media icon links
      - Best bet is to add screen-reader-only text inside each link (adjacent to the SVG icons) with text like "soAndSo on Twitter" or similar.
    - [ ] Remove the `<br>`s from the bios - they are read out by screen readers as "blank". Instead, wrap each block of text in `<p>` tags.
    - [ ] Consider making the bio content container an ARIA live region **but only when a user manually activates a thumbnail** so that people using screen readers know that _something_ happened as a result of their interaction.
  - [ ] Thumbnails
    - [ ] Organize the thumbnails using a `<ul>` so that screen readers know exactly how many items there are, and where they are in the list.
    - [ ] Get the organizer names on the thumbnails to become visible on focus, not just on hover.
    - [ ] Use `<button>`s instead of `<a>`s for the thumbnail images. They initiate on-page actions rather than URL changes.
      - To get the URL path to update, use the [History API](https://developer.mozilla.org/en-US/docs/Web/API/History)
    - [ ] Indicate which `<button>` is currently active by placing `aria-current="true"` on it.
      - It'd be awesome if this also triggered a visible style change too (like the hover/focus state) for sighted keyboard-based users.
    - [ ] Give each of the thumbnail `<img>`s an empty `alt`, since they are redundant with the adjacent overlay text. See WAI's [tutorial on Decorative Images](https://www.w3.org/WAI/tutorials/images/decorative/) for more.

### Code of Conduct modal dialog
  - [ ] When the modal opens, set focus on the close button. Without this, people using their keyboard or screen reader think nothing happened.
  - [ ] When the modal closes, return focus to the element that launched it (the "View Code of Conduct" button). This is so people using their keyboard or screen can "pick up" where they left off when they opened the modal.
  - [ ] Make it impossible to exit the modal when using your keyboard (using `Tab` and `Shift` + `Tab`). This is called "trapping" the keyboard focus.
    - When the user presses `Tab` while on the last focusable element, programmatically move focus to the first focusable element (and vice versa for `Shift` + `Tab`).
  - [ ] Add `role="dialog"`, `aria-modal="true"`, and `aria-labelledby` (pointing to the `[id]` on the heading that says "Code of Conduct").
  - [ ] Change the `<h4>` to an `<h1>`

### Footer
  - [ ] Change `<section id="contact">` to `<footer>` landmark tag (see WAI's [Page footer](https://www.w3.org/WAI/tutorials/page-structure/regions/#page-footer) notes).
  - [ ] Add missing form labels to input elements (ideally visually, associated with `for` attribute, but an `aria-label` could also work here). See [WCAG 2.0 3.3.2: Labels or Instructions](https://www.w3.org/TR/UNDERSTANDING-WCAG20/minimize-error-cues.html) for more.
    - [ ] Mailing list > email field with `id="mce-EMAIL"`
    - [ ] Mailing list > hidden text field inside `#mc_hidden`. 
      - Obviously not _really_ needed for humans because, well, the field is hidden. But to make some validators happy we can probably just slap an `aria-label` on this.

**Comment by jasonwebb** [comment]
FYI - I've started putting together some changes in feature branches on a personal fork of this repo, and I've added my name next to items in the task list to "claim" them. You can expect incoming PRs shortly! I'll drop into the Slack space to discuss reviewers and such unless anyone wants to comment here.

Right now I'm thinking that one feature branch (and then one PR) per "section" of tasks makes sense. For example, one branch/PR for all the "main navigation bar" items. If anyone has other ideas, I'm all ears!

If you'd like to work on any items (or are already), please drop a note here and I'll add your name next to them in the list!
